File: 2. Medium/15. 3Sum/MapSolution.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList();
        Map<Integer, Integer> dups = new HashMap();

        if (nums.length < 3)
            return res;

        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i++) {
            for (int j = i + 1; j < nums.length - 1; j++) {
                int sum2 = nums[i] + nums[j];

                Integer dup = dups.get(sum2);
                if (dup != null && dup == nums[i])
                    continue;

                int x = Arrays.binarySearch(nums, j + 1, nums.length, -sum2);
                if (x >= 0) {
                    dups.put(sum2, nums[i]);
                    res.add(Arrays.asList(nums[i], nums[j], nums[x]));
                }
            }
        }

        return res;
    }
}

File: 2. Medium/560. Subarray Sum Equals K/map-solution.ts
function subarraySum(nums: number[], k: number): number {
  let map = new Map<number, number>([[0, 1]]),
      res = 0,
      sum = 0;

  for (let i = 0; i < nums.length; ++i) {
    sum += nums[i];
    res += map.get(sum - k) ?? 0;
    map.set(sum, 1 + (map.get(sum) ?? 0));
  }

  return res;
}
